### PR TITLE
Handle JSON marshal error type into err.Error()

### LINKED ISSUE #8: JSON marshal does not work on error type
JSON doesn't properly marshal error types, causing the logger to essentially print out an empty JSON object whenever an `error` is part of the arg value. Potentially related GH issue: https://github.com/golang/go/issues/10748

Example:
```go
package main

import (
	"errors"
	"os"

	hclog "github.com/hashicorp/go-hclog"
)

func main() {
	err := errors.New("this is an error")
	logger := hclog.New(&hclog.LoggerOptions{
		Level:      hclog.Info,
		Output:     os.Stdout,
		JSONFormat: true,
	})

	logger.Info("hello from logger", "err", err)
}
```

Output:
```
{"@level":"info","@message":"hello from logger","@timestamp":"2017-08-26T22:02:31.816599-04:00","err":{}}
```

### PR DESCRIPTION
**Pull request by calvn**
Fixes #8

### CODE REVIEW

**Comment by jefferai** [comment]
The issue linked from #8 makes a pretty strong case that this is not the right behavior.

**Comment by calvn** [comment]
What should be the correct behavior? I took a look at how other loggers such as logrus handle this case, and they also use type assertion + `err.Error()` for encoding `error` types into JSON. https://github.com/sirupsen/logrus/issues/377

Another example is [apex/log](https://github.com/apex/log/blob/master/entry.go#L55), where the `WithError()` method calls `err.Error()` so that its JSON handler can properly encode the error.

**Comment by jefferai** [comment]
At a minimum we should probably check if the arg satisfies the json marshaler or textmarshaler interfaces before falling back on the plain string.

Arguably this should be done for any type. @evanphx thoughts?

**Comment by jefferai** [comment]
It seems @evanphx is on jury duty but hopefully he can get to looking at this soon.

File: logger_test.go [hunk 1]
```diff
@@ -103,10 +106,132 @@ func TestLogger(t *testing.T) {
...
+}
+
+// error impl.
+func (c *customErrText) Error() string {
+	return c.Message
+}
+
+// text.Marshaler impl.
+func (c customErrText) MarshalText() ([]byte, error) {
+	return []byte(c.Message), nil
```
Comment (jefferai) [inline logger_test.go]: If the Error function and the marshalling function return the same text it's hard to know which is actually being called in the test

**Comment by jefferai** [comment]
LGTM. @evanphx @ryanuber @mitchellh who will win the race to be the second reviewer?

**Pull request merged**